cpp/ixtheo_notation_stats.cc

```diff
@@ -0,0 +1,157 @@
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include <memory>
+#include <unordered_map>
+#include <cctype>
+#include <cstdlib>
+#include "DirectoryEntry.h"
+#include "Leader.h"
+#include "MarcUtil.h"
+#include "Subfields.h"
+#include "util.h"
+
+
+void Usage() {
+    std::cerr << "Usage: " << progname << "marc_input code_to_description_map\n";
+    std::exit(EXIT_FAILURE);
+}
+
+
+void LoadCodeToDescriptionMap(const std::shared_ptr<FILE> &code_to_description_map_file,
+			      const std::string &code_to_description_map_filename,
+			      std::unordered_map<std::string, std::string> * const code_to_description_map)
+{
+    char line[1024];
+    unsigned line_no(0);
+    while (std::fgets(line, sizeof line, code_to_description_map_file.get())) {
+	++line_no;
+	const size_t line_length(std::strlen(line));
+	if (line_length < 7) // 4 double quotes + 1 comma + two strings of length 1
+	    continue;
+
+	char *end(line + line_length);
+	char *quote(std::find(line + 1, end, '"'));
+	if (quote == end)
+	    Error("malformed line " + std::to_string(line_no) + " in \"" + code_to_description_map_filename
+		  + "\" (1)!");
+	*quote = '\0';
+	const std::string code(line + 1);
+	const char *description_start(quote + 2);
+	if (*description_start != '"')
+	    Error("malformed line " + std::to_string(line_no) + " in \"" + code_to_description_map_filename
+		  + "\" (2)!");
+	if (line[line_length - 2] != '"')
+	    Error("malformed line " + std::to_string(line_no) + " in \"" + code_to_description_map_filename
+		  + "\" (3)!");
+	line[line_length - 2] = '\0';
+	const std::string  description(description_start + 1);
+	(*code_to_description_map)[code] = description;
+    }
+
+    std::cerr << "Found " << code_to_description_map->size() << " code to description mappings.\n";
+}
+
+
+void CollectCounts(const std::shared_ptr<FILE> &input,
+		   const std::unordered_map<std::string, std::string> &code_to_description_map,
+		   std::unordered_map<std::string, unsigned> * const categories_to_counts_map)
+{
+    categories_to_counts_map->clear();
+
+    std::shared_ptr<Leader> leader;
+    std::vector<DirectoryEntry> dir_entries;
+    std::vector<std::string> field_data;
+    unsigned count(0), ixtheo_notation_count(0), records_with_ixtheo_notations(0);
+
+    std::string err_msg;
+    while (MarcUtil::ReadNextRecord(input.get(), leader, &dir_entries, &field_data, &err_msg)) {
+        ++count;
+
+        if (dir_entries[0].getTag() != "001")
+            Error("First field is not \"001\"!");
+
+	int lok_index(MarcUtil::GetFieldIndex(dir_entries, "LOK"));
+	bool found_at_least_one(false);
+	while (lok_index != -1) {
+	    const Subfields subfields(field_data[lok_index]);
+	    const auto begin_end(subfields.getIterators('0'));
+	    if (std::distance(begin_end.first, begin_end.second) == 2 and begin_end.first->second == "936ln") {
```

Comment:
You should use `begin_end.end` here. Even thoug you get the result of std::distance` right, it dosn't check if there are only two elements.